frontend/src/poc-code/testK8s/ws/WSFactory.ts

```diff
@@ -0,0 +1,338 @@
+import type {
+  EventHandlers,
+  EventHandlerTypes,
+  CloseHandler,
+  MessageHandler,
+  BulkMessageHandler,
+  MessageDataType,
+  DestroyHandler,
+  OpenHandler,
+  ErrorHandler,
+} from './types';
+import { applyConfigSubProtocols, applyConfigHost, createURL } from './ws-utils';
+
+export type WSOptions = {
+  /**
+   * The path to the resource you wish to watch.
+   */
+  path: string;
+  /**
+   * Overridable ws host url for Plugins. Normally set by the application.
+   */
+  host?: string;
+  /**
+   * Overridable ws sub protocols for Plugins. Normally set by the application.
+   * Is ignored if `host` is not set.
+   */
+  subProtocols?: string[];
+  // TODO: document
+  reconnect?: boolean;
+  jsonParse?: boolean;
+  bufferMax?: number;
+  bufferFlushInterval?: number;
+  timeout?: number;
+};
+
+/**
+ * @class WebSocket factory and utility wrapper.
+ */
+export class WSFactory {
+  private readonly handlers: EventHandlers;
+
+  private readonly flushCanceler: ReturnType<typeof setInterval>;
+
+  private readonly bufferMax: number;
+
+  private paused = false;
+
+  private state = '';
+
+  private messageBuffer: MessageDataType[] = [];
+
+  private connectionAttempt: ReturnType<typeof setTimeout>;
+
+  private ws: WebSocket | null = null;
+
+  /**
+   * @param id - unique id for the WebSocket.
+   * @param options - websocket options to initate the WebSocket with.
+   */
+  constructor(private readonly id: string, private readonly options: WSOptions) {
+    this.bufferMax = options.bufferMax || 0;
+    this.handlers = {
+      open: [],
+      close: [],
+      error: [],
+      message: [],
+      destroy: [],
+      bulkMessage: [],
+    };
+    this.connect();
+
+    if (this.bufferMax) {
+      this.flushCanceler = setInterval(
+        this.flushMessageBuffer.bind(this),
+        this.options.bufferFlushInterval || 500,
+      );
+    }
+  }
+
+  private reconnect() {
+    if (this.connectionAttempt || this.state === 'destroyed') {
+      return;
+    }
+
+    let delay = 1000;
+
+    const attempt = () => {
+      if (!this.options.reconnect || this.state === 'open') {
+        clearTimeout(this.connectionAttempt);
+        this.connectionAttempt = null;
+        return;
+      }
+      if (this.options.timeout && delay > this.options.timeout) {
+        clearTimeout(this.connectionAttempt);
+        this.connectionAttempt = null;
+        this.destroy();
+        return;
+      }
+
+      this.connect();
+      delay = Math.round(Math.min(1.5 * delay, 60000));
+      this.connectionAttempt = setTimeout(attempt, delay);
+      // eslint-disable-next-line no-console
+      console.log(`attempting reconnect in ${delay / 1000} seconds...`);
```

Comment:
Do we want to keep these messages in console?

Reply:
Nothing in `ws` folder is HAC-Core code, it's a shim for my work not being in the SDK NPM Package yet. Please take that to my WS PR (https://github.com/openshift/dynamic-plugin-sdk/pull/38) if you have a problem with excess echoing. `log`s fwiw are muted in Production.